nibabel/spatialimages.py

```diff
@@ -461,8 +479,66 @@ def from_image(klass, img):
                      klass.header_class.from_header(img.header),
                      extra=img.extra.copy())
 
+    def slice_affine(self, idx):
+        """ Retrieve affine for current image, if sliced by a given index
+
+        Applies scaling if down-sampling is applied, and adjusts the intercept
+        to account for any cropping.
+
+        Parameters
+        ----------
+        idx : numpy-compatible slice index
+
+        Returns
+        -------
+        affine : (4,4) ndarray
+            Affine with updated scale and intercept
+        """
+        idx = canonical_slicers(idx, self.shape, check_inds=False)[:3]
+
+        # Transform:
+        # sx  0  0 tx
+        #  0 sy  0 ty
+        #  0  0 sz tz
+        #  0  0  0  1
+        transform = np.eye(4, dtype=int)
+
+        for i, slicer in enumerate(idx):
+            if isinstance(slicer, slice):
+                if slicer.step == 0:
+                    raise ValueError("slice step cannot be 0")
+                transform[i, i] = slicer.step if slicer.step is not None else 1
+                transform[i, 3] = slicer.start or 0
+            elif isinstance(slicer, int):
+                transform[i, 3] = slicer
```

Comment:
Here I'm allowing the possibility that someone might use `slice_affine` with integer indices, even though I don't use it here.

Reply:
Right - the transform you're compiling is the transform from voxels in the cropped / subsampled array to voxels in the original array.   To get the final affine you apply the original affine to this transformation.